### DISCUSSION TITLE
Using concatenate dolar before parentheses

### DISCUSSION THREAD

**Discussion by jcubic** [peggyjs/peggy#578] [Q&A] [answered]
I have issue with this grammar:

```
content = text:$(image  / inline_text / list_item)+ nl:[\n]? {
    return text + (nl || '');
}
```

This doesn't work and don't give any error. It works when the group is in separated rule.

You can test the grammar on this CodePen demo:

https://codepen.io/jcubic/pen/zYLvPQa?editors=1000

If you add $ and replace text.join('') with text, the parser just output input text.

**Comment by hildjj** [comment]
This works as designed at the moment.  Changing it would break existing grammars.

This comes up most often for me in grammars that decode things like escaped quotes:

```peg.js
string 
  = DQUOTE s:char* DQUOTE { 
    return s.join('');
  } 
  
DQUOTE 
  = '"'
  
char 
  = ![\\"] @. // Anything but backslash or dquote
  / '\\' @'"'
```

The docs say:

> $ expression
> 
>     Try to match the expression. If the match succeeds, return the matched 
>     text instead of the match result.
> 
>     If you need to return the matched text in an action, you can use the 
>     text() function, but returning a labeled $ expression is sometimes more clear.

Which makes it pretty clear that `$` is doing the same thing as `text()`, but I'm very open to making this more explicit or clear if anyone wants to suggest more or different text.

**Reply by jcubic** [reply]
But why this works:

```
list_item = _ '*' __ text:$inline_text* [\n] {
  return '* ' + text + '\n';
}
```

It looks like `$` only works with `*` and not with `+`. If I change

```
inline_conent = image  / inline_text / list_item
content = text:$inline_conent* nl:[\n]? {
    return text + (nl || '');
}
```

I got an error:

> Parser: Possible infinite loop when parsing (repetition used with an expression that may not consume any input)

**Reply by hildjj** [reply]
If you remove that `$`, you should get the same error, which occurs because you've written a set of rules that resolves to something like `("b"*)+`.  If you try to match "c" with that, the part in parens always matches the empty string, and you keep iterating over that in an infinite loop trying to satisfy the `+`.

**Reply by jcubic** [reply]
Ok, but why `text:$inline_conent+` doesn't work and don't throw any error that ignore the input? Similar code works with `list_item`.

The inline_text is looks the same:

```
inline_text = bold / italic / link / [^\n]
```

For me it looks like a bug.

**Reply by hildjj** [reply]
I assume you mean this?

```peg.js
content = @$(inline_content+) nl:[\n]? 

inline_content = image  / inline_text / list_item
```

which works for me, I think?

**Reply by jcubic** [reply]
Can you explain why this works fine `text:$inline_text*`?

**Reply by hildjj** [reply]
Sorry, I've lost the thread a little bit while I was out running errands.  Can you tell me what you expected to happen and what actually happened, please?

**Reply by jcubic** [reply]
Ok I expect that this do work:

```peg
inline_content = image  / inline_text / list_item
    
content = text:$inline_content+ nl:[\n]? {
    return text + (nl || '');
}
```

same as:

```peg
inline_text = bold / italic / link / [^\n]

header = '#'+ ' '* text:$inline_text* [\n] {
    return '[[b;;]' + text + ']';
}
```

But the first grammar makes the parser ignore all the rules.

Without dollar, it works fine, but I need to use `text.join('')`. I don't see what is the different between those two rules.

**Reply by hildjj** [reply]
Ah, I think I see where your mental model doesn't match what Peggy is doing.  The `$` does not operate on the *results* of the rules that are returned.  It returns the slice of text from the original string that matched the rules.

**Reply by jcubic** [reply]
I'm not sure if I understand.

**Reply by hildjj** [reply, marked as answer]
Here's a simplified example:

```peg.js
Foo = bar+

bar = "bar" { return "baz" }
```

The input "barbar" will give the output `['baz', 'baz']`.

When you change Foo to be `$bar+`, the output is `"barbar"`.  The `$` discards whatever the `bar` rule returns and instead returns matching text.  The `bar` rule still runs twice, still matches twice.

**Reply by jcubic** [reply]
Ok, now I understand, it just that I used markdown header and list wrong in my input code,

If I use:
```
## [xxx](https://example.com)
```
It works the same, and the code land in output.
